feat: add signs to the editor and hunk management

Vinzent03 opened the pull request:
- Add support for showing hunks for unstaged changes (no staged changes as sign currently though) close #481 

<img width="260" height="263" alt="image" src="https://github.com/user-attachments/assets/f6ba7302-8ee0-4bc9-8990-f59620c28df3" />

- Stage\unstage individual hunks close #790

<img width="483" height="353" alt="image" src="https://github.com/user-attachments/assets/d9cf74ef-6760-4973-9cd9-ecb16ec11225" />


<img width="625" height="284" alt="image" src="https://github.com/user-attachments/assets/a3b0a3b9-1cb1-4144-a6ee-7f9c791e426c" />

Vinzent03 committed:
start with signs

Vinzent03 committed:
fixes and restructure

Vinzent03 committed:
use diff-match-patch and staged signs

Vinzent03 committed:
feat: hunk actions

Vinzent03 committed:
fix: properly manage extensions

Vinzent03 committed:
feat: stage hunk

Vinzent03 committed:
fix: small fixes

Vinzent03 committed:
refactor: better sign settings

Vinzent03 committed:
feat: display hunk changes inline

Vinzent03 committed:
feat: buttons for hunk actions and display on click

Vinzent03 committed:
fix: select hunks and unstage no_nl_at_eof

Vinzent03 committed:
fix: properly show tooltip

Vinzent03 committed:
feat: command to preview hunk

Vinzent03 committed:
fix: disable staged hunks as their computation is not feasible

Vinzent03 committed:
chore: fix eslint issues

Vinzent03 committed:
feat: stage individual hunks from split diff view

Vinzent03 committed:
refactor: use codemirror merge diff implementation

Vinzent03 committed:
fix: show tooltip for topdelete

Vinzent03 committed:
fix: don't remove dom elements in line authoring

Vinzent03 committed:
fix: make diff buttons horizontal

Vinzent03 committed:
chore: remove obsidian-community-lib package

Vinzent03 committed:
fix: reset new empty line

Vinzent03 committed:
fix: update settings description

Vinzent03 commented:
In case someone wants to try it out: here is an already built main.js and styles.css
[main.js](https://github.com/user-attachments/files/23136851/main.js)
[styles.css](https://github.com/user-attachments/files/23136852/styles.css)

(Or of course build it yourself)

chrisgrieser commented:
will have to see whether problems occur in everyday use, but so far, works great! Maybe the signs could be a bit thicker?

Given the size (2700+ lines) and the huge number of feature obsidian-git is accumulating, I feel like the git signs could rather be its own plugin, also since people could want the gitsigns without the need for the rest of the plugin. But I guess making use of the shared code functionality is probably an argument for making it one plugin.

chrisgrieser commented:
> It might look a bit strange in your case, but the signs gutter itself is pretty narrow. The huge gab is caused by the gutter margin in general. You would see the same gab when you have line numbers enabled.

true, but most people have line numbers disabled, and have no otther gutter plugins, resulting in this that weird spacing.

I tried this css snippet, and the alignment looks *much* nicer:

```css
.cm-gutters.cm-gutters-before:not(:has(> .cm-lineNumbers)) {
	margin-inline-end: 0;
	.signs-gutter-container {
		margin-inline-start: -1rem;
	}
}
```

<table border="1">
  <tr>
    <th>without signs before</th>
    <th>with signs before</th>
  </tr>
  <tr>
    <td><img width=60% alt="without signs before" src="https://github.com/user-attachments/assets/e2b15432-90a0-4d29-99d7-3a53e82ee9eb" /></td>
    <td><img width=60% alt="with signs before" src="https://github.com/user-attachments/assets/35e0b1de-e539-41ac-b785-1636f1eb8589" /></td>
  </tr>
</table>

<table border="1">
  <tr>
    <th>without signs after</th>
    <th>with signs after</th>
  </tr>
  <tr>
    <td><img width=60% alt="without signs after" src="https://github.com/user-attachments/assets/6b5c098b-e354-4aa3-b6e6-4a0480a78576" /></td>
    <td><img width=60% alt="with signs after" src="https://github.com/user-attachments/assets/89096588-9dd9-47a4-aa3d-6f687e3f6905" /></td>
  </tr>
</table>

chrisgrieser commented:
Another suggestion: a statusline item for the number of changes.

Not a game changer, but imo nice-to-have is an indicator for the number of changes in the current file (+ for additions, ~ for changes, – for deletions). Here is what it looks like in my editor:

<img width="288" height="60" alt="Screenshot_2025-10-26_09-46-07" src="https://github.com/user-attachments/assets/a1eaca15-32c6-4a3a-8863-b87160e71d5e" />

Even though the coloring is useful, the statusbar items in obsidian are usually uncolored, thus the decision whether to display the changes with color or not should probably be a toggle on the user end.

Vinzent03 committed:
fix: update styling

Vinzent03 committed:
feat: go to prev/next hunk commands

Vinzent03 committed:
feat: granular settings and change status bar

Vinzent03 committed:
fix: reset changing an empty line

Vinzent03 committed:
fix: properly disable signs via settings

Vinzent03 commented:
Thanks for your amount of suggestions and opinion!
- I've now altered the margin of the gutter if signs is the only gutter.
- I'm using the next hunk command all the time in Neovim as well, so I added it as well
- The status bar is a great idea. I've also added the option for it to be with or without color.

To the staging outside of Obsidian: Do you some the following settings enabled?
<img width="794" height="169" alt="image" src="https://github.com/user-attachments/assets/e096d2be-98b1-46ce-8fb3-d049430c423e" />
With these, as soon as you edit a file, it should refresh the cached index version. So it currently depends on the refresh of the source control view, which I get is not optimal.

Vinzent03 committed:
fix: use rem instead of px for dimensions

Vinzent03 committed:
fix: only show non 0 numbers in change status bar

Vinzent03 committed:
fix: refresh cached file index version every 10 seconds

This allows to be independent of the automatic refresh for the source
control view for when changes are staged outside of Obsidian.

chrisgrieser commented:
Thanks, looking good!

As of right now, I cannot think of any issues or much to improve. Only potential issue is that with a larger amount of hunks, Obsidian feels slightly laggy, but not sure if that's related to this plugin or not.

Vinzent03 committed:
fix: fallback to debounced diff for slow diffs

Vinzent03 committed:
fix: show signs for new hunks on refresh

Vinzent03 committed:
refactor: rename files and remove unused code

chrisgrieser commented:
awesome, thank you! looking forward to the next release!
